app/helpers/spree/admin/base_helper.rb

```diff
@@ -94,10 +105,11 @@ def preference_field_options(options)
       def preference_fields(object, form)
         return unless object.respond_to?(:preferences)
 
-        object.preferences.keys.map{ |key|
-          form.label("preferred_#{key}", Spree.t(key) + ": ") +
-            preference_field_for(form, "preferred_#{key}", type: object.preference_type(key))
-        }.join("<br />").html_safe
+        object.preferences.keys.map { |key|
+          [form.label("preferred_#{key}", Spree.t(key.to_s.gsub("_from_list", "")) + ": "),
+            preference_field_for(form, "preferred_#{key}",
+                                 { type: object.preference_type(key) }, object)]
+        }
```

Comment:
1. by putting this into an array you need to adapt the other usages of preference_fields, right? there are 3 other usages in other views, I think. And you need to add the magic join("<br />").html_safe

2. this is a bit too complicated, maybe use variables to make it easier to read:

```ruby
object.preferences.keys.map { |key|
      preference_label = form.label("preferred_#{key}", Spree.t(key.to_s.gsub("_from_list", "")) + ": ")
      preference_field = preference_field_for(form, "preferred_#{key}", { type: object.preference_type(key) }, object)
      [preference_label, preference_field]
```

Reply:
Good idea to use the variables to make it easier to read. 

I already did the adaption in the views (see [https://github.com/openfoodfoundation/openfoodnetwork/pull/5998/commits/d32ed6b48f1cffd68e008e7b12fcca0fd0c1d00b#diff-c543717515a0a6c36cfd7fb6402253e3R10-R13](here) for example. But I do now see that we could move the .html_safe back to the helper. 

The usage of [0] and [1] isn't great either. What about returning a hash, like `{ :label => "<label>...", { :field => "<select>..." }` Then the view could read `preference_fields[:label]`, for example.

Reply:
yes, with hashes it is better :+1: